src/common/buffer/Buffer.ts

```diff
@@ -242,13 +247,42 @@ export class Buffer implements IBuffer {
       // Trim the end of the line off if cols shrunk
       if (this._cols > newCols) {
         for (let i = 0; i < this.lines.length; i++) {
-          this.lines.get(i)!.resize(newCols, nullCell);
+          // +boolean for fast 0 or 1 conversion
+          dirtyMemoryLines += +this.lines.get(i)!.resize(newCols, nullCell);
         }
       }
     }
 
     this._cols = newCols;
     this._rows = newRows;
+
+    this._memoryCleanupQueue.clear();
+    // schedule memory cleanup only, if more than 10% of the lines are affected
+    if (dirtyMemoryLines > 0.1 * this.lines.length) {
+      this._memoryCleanupQueue.enqueue(() => this._batchedMemoryCleanup());
+    }
+  }
+
+  private _memoryCleanupQueue = new IdleTaskQueue();
+
+  private _batchedMemoryCleanup(): boolean {
+    let counted = 0;
+    for (let i = 0; i < this.lines.length; i++) {
+      counted += this.lines.get(i)!.cleanupMemory();
+      // throttle to 5k lines at once and
+      // return true to indicate, that the task is not finished yet
+      if (counted > 5000) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private _forceMemoryCleanup(): void {
+    this._memoryCleanupQueue.clear();
+    for (let i = 0; i < this.lines.length; i++) {
+      this.lines.get(i)!.cleanupMemory();
+    }
```

Comment:
Could this just be this?

```suggestion
    this._memoryCleanupQueue.flush();
```

Reply:
Or just remove it completely (not used by any logic atm)? The idea here was to provide a sync way to force cleanup the memory without the rescan overhead of multiple `_batchedMemoryCleanup` calls.